File: app/arrival_xml_parser_delegate.rb
class ArrivalXMLParserDelegate
  attr_reader :etas

  def initialize
    @etas = []
    @eta = nil
  end

  def parser(parser, didStartElement: name, namespaceURI: uri, qualifiedName: qual_name, attributes: attrs)
    @last = name

    if name == 'eta'
      @etas << @eta if @eta
      @eta = ETA.new
    end
  end

  def parser(parser, foundCharacters:string)
    case @last
    when 'destNm'
      @eta.dest_append(string)
    when 'rt'
      @eta.rt_append(string)
    when 'isApp'
      @eta.app_append(string)
    when 'arrT'
      @eta.arr_append(string)
    when 'prdt'
      @eta.prd_append(string)
    end
  end
end

File: Rail Line/ArrivalXMLParserDelegate.swift
import UIKit

class ArrivalXMLParserDelegate: UIViewController, XMLParserDelegate {
    private var last: String = ""
    private var eta: ETA?
    private var etas: [ETA] = []

    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String: String] = [:]) {
        last = elementName

        if last == "eta" {
            if eta != nil { etas.append(eta!) }
            eta = ETA()
        }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        switch last {
        case "destNm": eta?.destAppend(string)
        case "rt":     eta?.rtAppend(string)
        case "isApp":  eta?.appAppend(string)
        case "arrT":   eta?.arrAppend(string)
        case "prdt":   eta?.prdAppend(string)
        default:       break
        }
    }

    func getAllETAs() -> [ETA] {
        if let eta = eta { etas.append(eta) }
        eta = nil
        return etas
    }
}
